### PR TITLE
feat(docs): add ADR for Plugin lifecycle

### PR DESCRIPTION
**Pull request by IvoGoman**
# TL;DR;
This ADR proposes to introduce PluginDefinitionRevisions to enable version pinning on PluginPreset and Plugin resources. PluginPresets will support semantic version constraints, while Plugins, due to technical limitations, will only support concrete versions.
This will enable to stagger a rollout of a new PluginDefinition version by using a concrete version tag on the PluginPreset referring to this PluginDefinition. There is still the option to perform automatic upgrades as before.
Greenhouse will provide the means to use any given CD/GitOps/etc. way to deploy these manifests, but it will not provide an out of the box way to deploy the Greenhouse manifests of an Organization into it's namespace.
Garbage-collection of these revisions is proposed to delete the oldest unused revision/version over a count of `maxRevision`. This way gradually over time older revisions can be removed as soon they are no longer used.

### CODE REVIEW

File: architecture-decision-records/013-greenhouse-plugin-lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+- In the case of a bad update, there is no way to rollback to a previous version of the PluginDefinition. The Plugin is stuck in a broken state, depending on the issue and the HelmController a rollback to the previous state may or may not be possible. Either way the Plugin cannot be updated to the previous version of the PluginDefinition. Only a rollback of the whole PluginDefinition is possible, which will affect all Plugins using this PluginDefinition.
+
+To address these issues a new PluginDefinitionRevision CRD is proposed. The PluginDefinitionRevision should similar to a revision of a Helm release provide a snapshot of the PluginDefinition at a certain version. The PluginDefinitionRevision is immutable and stores the `version`, `pluginOptions` and `helmChart` of a PluginDefinition. There may be multiple revisions for a single PluginDefinition.
+Instead of referencing a PluginDefinition, a Plugin and PluginPreset should reference a PluginDefinitionRevision. This will allow to stage rollouts of new PluginDefinitions, as well as provide a way to rollback to a previous version of a PluginDefinition. Also updates to a PluginDefinition will not affect an in progress rollout.
```
Comment (ibakshay) [inline architecture-decision-records/013-greenhouse-plugin-lifecycle.md:361]: I like this approach a lot. The CNCF project `crossplane` also has the same behaviour and it works very well. Here is the [link to their documentation](https://docs.crossplane.io/latest/concepts/composition-revisions/#:~:text=Each%20time%20you%20edit%20a,the%20Composition%20-%20that%20unique%20state.).

> [!NOTE]
>When Composition Revisions are enabled three things happen:
Crossplane creates a CompositionRevision for each Composition update.
Composite Resources gain a spec.compositionRevisionRef field that specifies which CompositionRevision they use.
Composite Resources gain a spec.compositionUpdatePolicy field that specifies how they should be updated to new Composition Revisions.
> - Automatic: Automatically use the latest CompositionRevision. (Default)
> - Manual: Require manual intervention to change CompositionRevision.

Each time you edit a Composition Crossplane will automatically create a CompositionRevision that represents that ‘revision’ of the Composition - that unique state. Each revision is allocated an increasing revision number. This gives CompositionRevision consumers an idea about which revision is ’newest’. 

We can do the same with `PluginDefinition` and `PluginDefinitionRevision` :)

File: architecture-decision-records/013-greenhouse-plugin-lifecycle.md [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+spec:
+  clusterSelector: {...}
+  plugin:
+    ...
+    optionValues: {}
+    pluginDefinition:
+      name: cert-manager
+      version: "^v1.1.x"
+    releaseNamespace: kube-system
+```
```
Comment (ibakshay) [inline architecture-decision-records/013-greenhouse-plugin-lifecycle.md:421]: Here’s your refined GitHub review comment, keeping the arguments outside the manifest while also preserving the comments within the YAML for clarity:  

---

**How about introducing a `plugin-preset.spec.pluginDefinitionRevisionRef` field?**  

This change would provide more control over PluginDefinition revisions by allowing users to choose between automatic and manual updates:  

- **Automatic (`updatePolicy: automatic`)**: The Greenhouse controller will update `pluginDefinitionRevisionRef.name` to the latest available PluginDefinitionRevision automatically.  
- **Manual (`updatePolicy: manual`)**: Users must manually update `pluginDefinitionRevisionRef.name` when they want to upgrade.  

This ensures that the teams can choose between fully automated rollouts vs manual process. 

```yaml
apiVersion: greenhouse.sap/v1alpha1
kind: PluginPreset
metadata:
  name: cert-manager
spec:
  clusterSelector: {...}
  plugin:
    ...
    optionValues: {}
    pluginDefinition:
      name: cert-manager
    # The "updatePolicy" determines whether the PluginPreset should automatically update 
    # to the latest PluginDefinitionRevision.
    updatePolicy: manual | automatic
    pluginDefinitionRevisionRef: 
      name: cert-manager-zf123 
    # If "updatePolicy" is set to "automatic," the Greenhouse controller will automatically
    # update `pluginDefinitionRevisionRef.name` to the latest PluginDefinitionRevision.
    # If set to "manual," users must manually update `pluginDefinitionRevisionRef.name` 
    # when they want to upgrade.
    releaseNamespace: kube-system
```

Would love to hear your thoughts on this! :)

Comment (ibakshay) [inline architecture-decision-records/013-greenhouse-plugin-lifecycle.md:421]: **How about introducing a `PluginPreset.spec.pluginDefinitionRevisionRef` field?**  

This change would provide more control over PluginDefinition revisions by allowing users to choose between automatic and manual updates:  

- **Automatic (`updatePolicy: automatic`)**: The Greenhouse controller will update `pluginDefinitionRevisionRef.name` to the latest available PluginDefinitionRevision automatically.  
- **Manual (`updatePolicy: manual`)**: Users must manually update `pluginDefinitionRevisionRef.name` when they want to upgrade.  

This ensures that the teams can choose between fully automated rollouts vs manual process. 

```yaml
apiVersion: greenhouse.sap/v1alpha1
kind: PluginPreset
metadata:
  name: cert-manager
spec:
  clusterSelector: {...}
  plugin:
    ...
    optionValues: {}
    pluginDefinition:
      name: cert-manager
    # The "updatePolicy" determines whether the PluginPreset should automatically update 
    # to the latest PluginDefinitionRevision.
    updatePolicy: manual | automatic
    pluginDefinitionRevisionRef: 
      name: cert-manager-zf123 
    # If "updatePolicy" is set to "automatic," the Greenhouse controller will automatically
    # update `pluginDefinitionRevisionRef.name` to the latest PluginDefinitionRevision.
    # If set to "manual," users must manually update `pluginDefinitionRevisionRef.name` 
    # when they want to upgrade.
    releaseNamespace: kube-system
```

Would love to hear your thoughts on this! :)

**Comment by abhijith-darshan** [comment]
as a first proposal for the PluginCatalog we could do -

```yaml
apiVersion: greenhouse.sap/v1alpha1
kind: PluginCatalog
metadata:
  name: catalog
  namespace: <org-name>
spec:
  sources:
    - github:
        repoURL: https://github.com/cloudoperators/greenhouse
        secretRef:
          name: github-token
          key: github
          mapping:
            token: token
            appID: appID
            privateKey: privateKey
        targetRevision: main
        path: perses/plugindefinition.yaml # (Explicit path to plugin definition file or can it be directory containing plugin definitions and controller parses all yamls in the base of the dir and extracts all plugin definitions?)
        paths:
          - perses/plugindefinition.yaml
          - alerts/plugindefinition.yaml
          - exposed-services/plugindefinition.yaml
        syncOptions:
          syncInterval: 1h # (How often to sync the catalog with the source)
          automated: true # (Automatically sync the catalog with the source) (mutually exclusive with syncInterval - if syncInterval is set, automated is ignored)
          prune: true # (Remove plugindefinitions that are not in the source) (default: false)


## do we need caching of the catalog? use redis / valkey like other open source projects?
  ## benefits of caching: avoid rate limiting, faster response times, avoid network latency
## do one api call to github to get diff if no diff compare the current state with the cache, if there is a diff in current state then replace with cache

## Do we support buckets?? Many buckets (s3, GCS) support version number or timestamp or generation number, we can use that to check for sync
```
In target revision we can do -

**Pinning Strategies** 
- branch name, 
- commit hash, 
- v tag

**Watch Strategies** on tags or releases

- v1.x.x (depending on the desired revision count we bring the definitions in descending order)

Catalog can create PluginDefinitions suffixed with revision...

**Comment by IvoGoman** [comment]
> as a first proposal for the PluginCatalog we could do -
> 
> ```yaml
> apiVersion: greenhouse.sap/v1alpha1
> kind: PluginCatalog
> metadata:
>   name: catalog
>   namespace: <org-name>
> spec:
>   sources:
>     - github:
>         repoURL: https://github.com/cloudoperators/greenhouse
>         secretRef:
>           name: github-token
>           key: github
>           mapping:
>             token: token
>             appID: appID
>             privateKey: privateKey
>         targetRevision: main
>         path: perses/plugindefinition.yaml # (Explicit path to plugin definition file or can it be directory containing plugin definitions and controller parses all yamls in the base of the dir and extracts all plugin definitions?)
>         paths:
>           - perses/plugindefinition.yaml
>           - alerts/plugindefinition.yaml
>           - exposed-services/plugindefinition.yaml
>         syncOptions:
>           syncInterval: 1h # (How often to sync the catalog with the source)
>           automated: true # (Automatically sync the catalog with the source) (mutually exclusive with syncInterval - if syncInterval is set, automated is ignored)
>           prune: true # (Remove plugindefinitions that are not in the source) (default: false)
> 
> ## Do we support buckets?? Many buckets (s3, GCS) support version number or timestamp or generation number, we can use that to check for sync
> ```
> 
> In target revision we can do -
> 
> **Pinning Strategies**
> 
> * branch name,
> * commit hash,
> * v tag
> 
> **Watch Strategies** on tags or releases
> 
> * v1.x.x

This PR is concerning the lifecycle, could you add your comment to #20 instead?

File: architecture-decision-records/013-greenhouse-plugin-lifecycle.md (hunk 2 above)
Comment (IvoGoman) [reply architecture-decision-records/013-greenhouse-plugin-lifecycle.md]: With referring explicit versions it is possible to use semantic version constraints to have automatic updates on minor/patch version upgrades etc.. These versions could also be used with Dependabot to create version update PRs on PluginPresets/Plugins. 
It would also be possible to stop automatic updates if the underlying chart receives an major version update.

Using the RevisionRef it will restrict to forward-fix only, no patches possible to previous minor versions. To ensure automatic updates applied in the correct order it is necessary to forbid updating PluginDefinitions to a lower HelmChart version than currently existing in the cluster.
This would automatically perform all updates, also for major versions.

| PluginDefinition version | RevisionNumber | RevisionRef |
| -- | -- | -- |
| v1.0.1 | 1 | cert-manager-zf123 |
| v2.0.1 | 2 | cert-manager-zf345 | 
| v1.0.3 | 3 | cert-manager-ab123 |

Using RevisionRefs to refer to a particular PluginDefinitionRevision will require manual updates in the PluginPreset/Plugin that are stored in a Git Repository, or in the UI. 
It would still be required to have the versions on the PluginDefinitions to be able to easily compare changes between Plugin versions.

**Review by abhijith-darshan — APPROVED** [review]

File: architecture-decision-records/013-greenhouse-plugin-lifecycle.md [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+
+| Decision Driver     | Rating | Reason                        |
+|---------------------|--------|-------------------------------|
+| [decision driver a] | +++    | Good, because [argument a]    |                                                                                                                                                                                                                                                                | 
+| [decision driver b] | ---    | Good, because [argument b]    |
+| [decision driver c] | --     | Bad, because [argument c]     |
+| [decision driver d] | o      | Neutral, because [argument d] |
+
+
```
[the remark this answers is not in the document]
Comment (IvoGoman) [reply architecture-decision-records/013-greenhouse-plugin-lifecycle.md]: This was resolved in a huddle
